### PR TITLE
[java] Fix FN in UnnecessaryModifierRule

### PR DESCRIPTION
**Pull request by jsotuyod**
- Enum constructors are inherently private, no need to add that
modifier

### CODE REVIEW

**Comment by oowekyala** [comment]
This is ok, I think it could be nice to enhance this rule by stating *which* modifier is redundant in the rule message. Do you want to do that yourself @jsotuyod ? Otherwise I'll do that when I merge

**Comment by jsotuyod** [comment]
@oowekyala feel free to do so. I don't think I'll have the chance to get to it before the weekend.

**Comment by oowekyala** [comment]
@jsotuyod I'm done improving the rule's messages. Please review and let me know if it's ok

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractAnyTypeDeclaration.java [hunk 1]
```diff
@@ -26,16 +27,52 @@
...
+    /**
+     * Returns the enclosing type of this type, if it is nested.
+     * Otherwise returns null. This won't consider anonymous classes
+     * until #905 is tackled. TODO 7.0.0
+     */
+    public final ASTAnyTypeDeclaration getEnclosingTypeDeclaration() {
+        if (!isNested()) {
+            return null;
+        }
+        return (ASTAnyTypeDeclaration) getNthParent(3);
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractAnyTypeDeclaration.java]: I'm a little concerned about this unchecked cast...

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractAnyTypeDeclaration.java [hunk 2]
```diff
@@ -26,16 +27,52 @@
...
+     * @param kinds Kinds to test
+     */
+    // TODO 7.0.0 move that up to ASTAnyTypeDeclaration
+    public final boolean enclosingTypeIsA(TypeKind... kinds) {
+
+        if (!isNested()) {
+            return false;
+        }
+
+        Node parent = getNthParent(3);
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractAnyTypeDeclaration.java]: should this be a call to `getEnclosingTypeDeclaration`?

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/UnnecessaryModifierRule.java [hunk 3]
```diff
@@ -32,94 +45,162 @@ public UnnecessaryModifierRule() {
...
+        return node instanceof ASTMethodDeclaration
+               ? ((ASTMethodDeclaration) node).getMethodName()
+               : node instanceof ASTMethodOrConstructorDeclaration
+                 // constructors are differentiated by their parameters, while we only use method name for methods
+                 ? ((ASTConstructorDeclaration) node).getQualifiedName().getOperation()
+                 : node instanceof ASTFieldDeclaration
+                   ? ((ASTFieldDeclaration) node).getVariableName()
+                   : node instanceof ASTResource
+                     ? ((ASTResource) node).getVariableDeclaratorId().getImage()
+                     : node.getImage();
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/UnnecessaryModifierRule.java]: I'd explode this into if statements as we do in `getPrintableNodeKind`... may be longer, but will enhance readability

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/UnnecessaryModifierRule.java [hunk 4]
```diff
@@ -130,57 +211,75 @@ public Object visit(final ASTMethodDeclaration node, Object data) {
...
-            }
-        }
-        
-        return false;
+        return node.isAnnotationPresent(SafeVarargs.class.getName());
     }
 
-    private void check(Node fieldOrMethod, Object data) {
+
+    private void checkDeclarationInInterfaceType(Object data, Node fieldOrMethod, List<Modifier> unnecessary) {
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/UnnecessaryModifierRule.java]: possibly better to use an `EnumSet<Modifier>` both for performance and avoiding duplicates

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractAnyTypeDeclaration.java (hunk 1 above)
Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractAnyTypeDeclaration.java]: Oh yeah you're right, I though isNested protected it but it doesn't check high enough in the tree

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractAnyTypeDeclaration.java [hunk 5]
```diff
@@ -70,7 +70,9 @@ public final ASTAnyTypeDeclaration getEnclosingTypeDeclaration() {
         if (!isNested()) {
             return null;
         }
-        return (ASTAnyTypeDeclaration) getNthParent(3);
+        Node parent = getNthParent(3);
+
+        return parent instanceof ASTAnyTypeDeclaration ? (ASTAnyTypeDeclaration) getNthParent(3) : null;
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractAnyTypeDeclaration.java]: `return parent instanceof ASTAnyTypeDeclaration ? (ASTAnyTypeDeclaration) parent : null;`

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractAnyTypeDeclaration.java [hunk 6]
```diff
@@ -49,10 +49,10 @@ public final boolean enclosingTypeIsA(TypeKind... kinds) {
             return false;
         }
 
-        Node parent = getNthParent(3);
+        ASTAnyTypeDeclaration parent = getEnclosingTypeDeclaration();
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractAnyTypeDeclaration.java]: missing a null-check here

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractAnyTypeDeclaration.java (hunk 5 above)
Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractAnyTypeDeclaration.java]: lol right

**Comment by jsotuyod** [comment]
@oowekyala thanks for your help, this is now merged.

**Pull request merged**